File: atlas-ai-py/atlas_ai/prediction/predictive_model.py
import math
import time
from copy import deepcopy


class PredictionMetrics:
    def __init__(self):
        self.mse: float = 0
        self.rmse: float = 0
        self.mae: float = 0
        self.r2: float = 0
        self.mape: float = 0


class TrainingHistory:
    def __init__(self):
        self.epoch: list[int] = []
        self.loss: list[float] = []
        self.val_loss: list[float] = []
        self.metrics: list[PredictionMetrics] = []


class ForecastResult:
    def __init__(self):
        self.predictions: list[float] = []
        self.confidence: list[float] = []
        self.horizon: int = 0
        self.timestamps: list[float] = []


class PredictiveModel:
    def __init__(self, window_size: int = 5, learning_rate: float = 0.01):
        self.model: list[float] = []
        self.history = TrainingHistory()
        self.metrics = PredictionMetrics()
        self.window_size = window_size
        self.learning_rate = learning_rate
        self.initialized = False

    def train(
        self,
        data: list[float],
        epochs: int = 100,
        validation_split: float = 0.2,
    ) -> TrainingHistory:
        if not self.initialized or len(self.model) == 0:
            self._initialize_model(data)

        train_size = int(len(data) * (1 - validation_split))
        train_data = data[:train_size]
        val_data = data[train_size:]

        for epoch in range(epochs):
            train_loss = 0.0
            train_samples = 0

            for i in range(len(train_data) - self.window_size):
                window = train_data[i:i + self.window_size]
                target = train_data[i + self.window_size]

                prediction = self._predict_step(window)
                error = target - prediction

                self._update_weights(window, error)

                train_loss += error * error
                train_samples += 1

            avg_train_loss = train_loss / max(1, train_samples)
            self.history.loss.append(avg_train_loss)
            self.history.epoch.append(epoch)

            val_loss = 0.0
            val_samples = 0

            for i in range(len(val_data) - self.window_size):
                window = val_data[i:i + self.window_size]
                target = val_data[i + self.window_size]
                prediction = self._predict_step(window)
                val_loss += (target - prediction) ** 2
                val_samples += 1

            avg_val_loss = val_loss / max(1, val_samples)
            self.history.val_loss.append(avg_val_loss)

            if epoch > 10 and avg_train_loss < 0.0001:
                break

        return self.history

    def predict(self, input_data: list[float]) -> ForecastResult:
        horizon = min(len(input_data) - self.window_size, 20)
        predictions: list[float] = []
        confidence: list[float] = []
        timestamps: list[float] = []

        for i in range(horizon):
            window = input_data[i:i + self.window_size]
            prediction = self._predict_step(window)
            predictions.append(prediction)

            variance = self._calculate_variance(window)
            conf = max(0.5, min(0.95, 1 - math.sqrt(variance) * 0.1))
            confidence.append(conf)

            timestamps.append(time.time() + i * 1000)

        result = ForecastResult()
        result.predictions = predictions
        result.confidence = confidence
        result.horizon = horizon
        result.timestamps = timestamps
        return result

    def forecast(self, input_data: list[float], steps: int) -> ForecastResult:
        predictions: list[float] = []
        confidence: list[float] = []
        timestamps: list[float] = []
        working_data = list(input_data)

        for i in range(steps):
            window = working_data[-self.window_size:]
            prediction = self._predict_step(window)
            predictions.append(prediction)
            working_data.append(prediction)

            conf = max(0.3, 0.9 - i * 0.05)
            confidence.append(conf)

            timestamps.append(time.time() + (len(input_data) + i) * 1000)

        result = ForecastResult()
        result.predictions = predictions
        result.confidence = confidence
        result.horizon = steps
        result.timestamps = timestamps
        return result

    def evaluate(self, test_data: list[float]) -> PredictionMetrics:
        if len(test_data) <= self.window_size:
            return self.metrics

        actual: list[float] = []
        predicted: list[float] = []

        for i in range(len(test_data) - self.window_size):
            window = test_data[i:i + self.window_size]
            target = test_data[i + self.window_size]
            actual.append(target)
            predicted.append(self._predict_step(window))

        n = len(actual)
        mse = sum((a - predicted[i]) ** 2 for i, a in enumerate(actual)) / n
        rmse = math.sqrt(mse)
        mae = sum(abs(a - predicted[i]) for i, a in enumerate(actual)) / n

        mean_actual = sum(actual) / n
        ss_total = sum((a - mean_actual) ** 2 for a in actual)
        ss_residual = sum((a - predicted[i]) ** 2 for i, a in enumerate(actual))
        r2 = 1 - ss_residual / (ss_total + 1e-10)

        mape = (
            sum(abs((a - predicted[i]) / (a + 1e-10)) for i, a in enumerate(actual))
            * 100
            / n
        )

        self.metrics.mse = mse if mse else 0
        self.metrics.rmse = rmse if rmse else 0
        self.metrics.mae = mae if mae else 0
        self.metrics.r2 = max(0, r2)
        self.metrics.mape = mape if mape else 0

        return self.metrics

    def get_history(self) -> TrainingHistory:
        return deepcopy(self.history)

    def get_metrics(self) -> PredictionMetrics:
        return deepcopy(self.metrics)

    def get_model(self) -> list[float]:
        return list(self.model)

    def set_model(self, weights: list[float]):
        self.model = list(weights)
        self.initialized = True

    def reset(self):
        self.model = []
        self.history = TrainingHistory()
        self.metrics = PredictionMetrics()
        self.initialized = False

    def to_json(self) -> dict:
        return {
            "model": self.model,
            "windowSize": self.window_size,
            "learningRate": self.learning_rate,
            "history": {
                "epoch": self.history.epoch,
                "loss": self.history.loss,
                "valLoss": self.history.val_loss,
                "metrics": [{"mse": m.mse, "rmse": m.rmse, "mae": m.mae, "r2": m.r2, "mape": m.mape} for m in self.history.metrics],
            },
            "metrics": {
                "mse": self.metrics.mse,
                "rmse": self.metrics.rmse,
                "mae": self.metrics.mae,
                "r2": self.metrics.r2,
                "mape": self.metrics.mape,
            },
            "initialized": self.initialized,
        }

    @staticmethod
    def from_json(data: dict):
        model = PredictiveModel(data.get("windowSize", 5), data.get("learningRate", 0.01))
        model.model = data.get("model", [])
        hist_data = data.get("history", {})
        model.history.epoch = hist_data.get("epoch", [])
        model.history.loss = hist_data.get("loss", [])
        model.history.val_loss = hist_data.get("valLoss", [])
        met_data = data.get("metrics", {})
        model.metrics.mse = met_data.get("mse", 0)
        model.metrics.rmse = met_data.get("rmse", 0)
        model.metrics.mae = met_data.get("mae", 0)
        model.metrics.r2 = met_data.get("r2", 0)
        model.metrics.mape = met_data.get("mape", 0)
        model.initialized = data.get("initialized", False)
        return model

    def _initialize_model(self, data: list[float]):
        n = self.window_size + 1
        self.model = [1.0 / n] * n
        self.initialized = True

    def _predict_step(self, window: list[float]) -> float:
        if len(self.model) == 0:
            return sum(window) / len(window) if window else 0

        prediction = 0.0
        for i in range(min(len(window), len(self.model))):
            prediction += window[i] * self.model[i]
        return prediction

    def _update_weights(self, window: list[float], error: float):
        gradient = self.learning_rate * error
        for i in range(min(len(window), len(self.model))):
            self.model[i] += gradient * window[i] * 0.01

    def _calculate_variance(self, data: list[float]) -> float:
        if not data:
            return 0
        mean = sum(data) / len(data)
        return sum((v - mean) ** 2 for v in data) / len(data)

File: atlas-ai_deprecated/Prediction/PredictiveModel.ts
/**
 * PredictiveModel - Time series forecasting and prediction model
 * Supports training, prediction, and evaluation metrics
 */

export interface PredictionMetrics {
  mse: number;
  rmse: number;
  mae: number;
  r2: number;
  mape: number;
}

export interface TrainingHistory {
  epoch: number[];
  loss: number[];
  valLoss: number[];
  metrics: PredictionMetrics[];
}

export interface ForecastResult {
  predictions: number[];
  confidence: number[];
  horizon: number;
  timestamps: number[];
}

/**
 * PredictiveModel - Forecasting model for time series and general prediction
 */
export class PredictiveModel {
  private model: number[];
  private history: TrainingHistory;
  private metrics: PredictionMetrics;
  private windowSize: number;
  private learningRate: number;
  private initialized: boolean;

  constructor(windowSize: number = 5, learningRate: number = 0.01) {
    this.model = [];
    this.history = {
      epoch: [],
      loss: [],
      valLoss: [],
      metrics: [],
    };
    this.metrics = {
      mse: 0,
      rmse: 0,
      mae: 0,
      r2: 0,
      mape: 0,
    };
    this.windowSize = windowSize;
    this.learningRate = learningRate;
    this.initialized = false;
  }

  /**
   * Trains the model on historical data
   */
  train(
    data: number[],
    epochs: number = 100,
    validationSplit: number = 0.2
  ): TrainingHistory {
    // Initialize model weights if not done
    if (!this.initialized || this.model.length === 0) {
      this.initializeModel(data);
    }

    const trainSize = Math.floor(data.length * (1 - validationSplit));
    const trainData = data.slice(0, trainSize);
    const valData = data.slice(trainSize);

    for (let epoch = 0; epoch < epochs; epoch++) {
      // Train on sliding windows
      let trainLoss = 0;
      let trainSamples = 0;

      for (let i = 0; i < trainData.length - this.windowSize; i++) {
        const window = trainData.slice(i, i + this.windowSize);
        const target = trainData[i + this.windowSize];

        const prediction = this.predictStep(window);
        const error = target - prediction;

        // Update model weights
        this.updateWeights(window, error);

        trainLoss += error * error;
        trainSamples++;
      }

      const avgTrainLoss = trainLoss / Math.max(1, trainSamples);
      this.history.loss.push(avgTrainLoss);
      this.history.epoch.push(epoch);

      // Validation loss
      let valLoss = 0;
      let valSamples = 0;

      for (let i = 0; i < valData.length - this.windowSize; i++) {
        const window = valData.slice(i, i + this.windowSize);
        const target = valData[i + this.windowSize];
        const prediction = this.predictStep(window);
        valLoss += Math.pow(target - prediction, 2);
        valSamples++;
      }

      const avgValLoss = valLoss / Math.max(1, valSamples);
      this.history.valLoss.push(avgValLoss);

      // Early stopping
      if (epoch > 10 && avgTrainLoss < 0.0001) {
        break;
      }
    }

    return this.history;
  }

  /**
   * Makes predictions on new data
   */
  predict(inputData: number[]): ForecastResult {
    const horizon = Math.min(inputData.length - this.windowSize, 20);
    const predictions: number[] = [];
    const confidence: number[] = [];
    const timestamps: number[] = [];

    for (let i = 0; i < horizon; i++) {
      const window = inputData.slice(i, i + this.windowSize);
      const prediction = this.predictStep(window);
      predictions.push(prediction);

      // Calculate confidence based on prediction variance
      const variance = this.calculateVariance(window);
      const conf = Math.max(0.5, Math.min(0.95, 1 - Math.sqrt(variance) * 0.1));
      confidence.push(conf);

      timestamps.push(Date.now() + i * 1000); // Assume 1 second intervals
    }

    return {
      predictions,
      confidence,
      horizon,
      timestamps,
    };
  }

  /**
   * Forecasts future values beyond the input data
   */
  forecast(inputData: number[], steps: number): ForecastResult {
    const predictions: number[] = [];
    const confidence: number[] = [];
    const timestamps: number[] = [];
    const workingData = [...inputData];

    for (let i = 0; i < steps; i++) {
      const window = workingData.slice(-this.windowSize);
      const prediction = this.predictStep(window);
      predictions.push(prediction);
      workingData.push(prediction);

      // Decrease confidence for longer horizons
      const conf = Math.max(0.3, 0.9 - i * 0.05);
      confidence.push(conf);

      timestamps.push(Date.now() + (inputData.length + i) * 1000);
    }

    return {
      predictions,
      confidence,
      horizon: steps,
      timestamps,
    };
  }

  /**
   * Evaluates the model on test data
   */
  evaluate(testData: number[]): PredictionMetrics {
    if (testData.length <= this.windowSize) {
      return this.metrics;
    }

    const actual: number[] = [];
    const predicted: number[] = [];

    for (let i = 0; i < testData.length - this.windowSize; i++) {
      const window = testData.slice(i, i + this.windowSize);
      const target = testData[i + this.windowSize];
      actual.push(target);
      predicted.push(this.predictStep(window));
    }

    // Calculate metrics
    const n = actual.length;
    const mse = actual.reduce((sum, a, i) => sum + Math.pow(a - predicted[i], 2), 0) / n;
    const rmse = Math.sqrt(mse);
    const mae = actual.reduce((sum, a, i) => sum + Math.abs(a - predicted[i]), 0) / n;

    // R-squared
    const meanActual = actual.reduce((sum, v) => sum + v, 0) / n;
    const ssTotal = actual.reduce((sum, a) => sum + Math.pow(a - meanActual, 2), 0);
    const ssResidual = actual.reduce((sum, a, i) => sum + Math.pow(a - predicted[i], 2), 0);
    const r2 = 1 - ssResidual / (ssTotal + 1e-10);

    // MAPE
    const mape =
      actual.reduce((sum, a, i) => sum + Math.abs((a - predicted[i]) / (a + 1e-10)), 0) * 100 / n;

    this.metrics = {
      mse: mse || 0,
      rmse: rmse || 0,
      mae: mae || 0,
      r2: Math.max(0, r2),
      mape: mape || 0,
    };

    return this.metrics;
  }

  /**
   * Gets the training history
   */
  getHistory(): TrainingHistory {
    return { ...this.history };
  }

  /**
   * Gets current metrics
   */
  getMetrics(): PredictionMetrics {
    return { ...this.metrics };
  }

  /**
   * Gets the model parameters
   */
  getModel(): number[] {
    return [...this.model];
  }

  /**
   * Sets the model parameters
   */
  setModel(weights: number[]): void {
    this.model = [...weights];
    this.initialized = true;
  }

  /**
   * Resets the model
   */
  reset(): void {
    this.model = [];
    this.history = {
      epoch: [],
      loss: [],
      valLoss: [],
      metrics: [],
    };
    this.metrics = {
      mse: 0,
      rmse: 0,
      mae: 0,
      r2: 0,
      mape: 0,
    };
    this.initialized = false;
  }

  /**
   * Serializes the model to JSON
   */
  toJSON(): object {
    return {
      model: this.model,
      windowSize: this.windowSize,
      learningRate: this.learningRate,
      history: this.history,
      metrics: this.metrics,
      initialized: this.initialized,
    };
  }

  /**
   * Creates a model from serialized data
   */
  static fromJSON(data: object): PredictiveModel {
    const json = data as {
      model: number[];
      windowSize: number;
      learningRate: number;
      history: TrainingHistory;
      metrics: PredictionMetrics;
      initialized: boolean;
    };
    const model = new PredictiveModel(json.windowSize, json.learningRate);
    model.model = json.model;
    model.history = json.history;
    model.metrics = json.metrics;
    model.initialized = json.initialized;
    return model;
  }

  /**
   * Initializes the model with optimal weights
   */
  private initializeModel(data: number[]): void {
    const n = this.windowSize + 1;
    this.model = new Array(n).fill(0).map((_, i) => {
      // Linear trend initialization
      return i === 0 ? 1 / n : 1 / n;
    });
    this.initialized = true;
  }

  /**
   * Single prediction step
   */
  private predictStep(window: number[]): number {
    if (this.model.length === 0) {
      // Simple average if model not trained
      return window.reduce((a, b) => a + b, 0) / window.length;
    }

    let prediction = 0;
    for (let i = 0; i < Math.min(window.length, this.model.length); i++) {
      prediction += window[i] * this.model[i];
    }

    return prediction;
  }

  /**
   * Updates model weights using gradient descent
   */
  private updateWeights(window: number[], error: number): void {
    const gradient = this.learningRate * error;

    for (let i = 0; i < Math.min(window.length, this.model.length); i++) {
      this.model[i] += gradient * window[i];
    }
  }

  /**
   * Calculates variance of a window
   */
  private calculateVariance(data: number[]): number {
    const mean = data.reduce((a, b) => a + b, 0) / data.length;
    return data.reduce((sum, v) => sum + Math.pow(v - mean, 2), 0) / data.length;
  }
}